### PR TITLE
Add support for V2 of AWS's instance metadata service (IMDS)

### PR DESCRIPTION
**Pull request by justincjohnson**

### CODE REVIEW

File: rebar.lock [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{"1.1.0",
+{"1.2.0",
```
Comment (justincjohnson) [inline rebar.lock:1]: Bumped when getting deps with rebar3.

File: src/aws_credentials_ec2.erl [hunk 2]
```diff
@@ -3,40 +3,70 @@
...
 
+-define(SESSION_TOKEN_URL,
+        "http://169.254.169.254/latest/api/token/").
+-define(SESSION_TOKEN_TTL_HEADER,
+        "x-aws-ec2-metadata-token-ttl-seconds").
+-define(SESSION_TOKEN_TTL_SECONDS,
+        "21600").
+-define(SESSION_TOKEN_HEADER,
+        "X-aws-ec2-metadata-token").
 -define(CREDENTIAL_URL,
```
Comment (justincjohnson) [inline src/aws_credentials_ec2.erl:14]: Note that the current changes on master are broken from the commit that changed the format of this macro.

File: src/aws_credentials_ec2.erl [hunk 3]
```diff
@@ -3,40 +3,70 @@
...
         "http://169.254.169.254/latest/meta-data/iam/security-credentials/").
 -define(DOCUMENT_URL,
         "http://169.254.169.254/latest/dynamic/instance-identity/document").
 
 -type role() :: binary().
+-type session_token() :: binary().
```
Comment (justincjohnson) [inline src/aws_credentials_ec2.erl:20]: I'm new to Erlang and feel like I'm possibly spec'ing too much?  🤷🏻  Please educate me.

File: src/aws_credentials_ec2.erl [hunk 4]
```diff
@@ -3,40 +3,70 @@
...
 -export([fetch/1]).
 
 -spec fetch(aws_credentials_provider:options()) ->
         {ok, aws_credentials_provider:credentials(), aws_credentials_provider:expiration()} |
         {error, any()}.
 fetch(_Options) ->
-    {ok, Role} = fetch_role(),
-    {ok, AccessKeyID, SecretAccessKey, ExpirationTime, Token} = fetch_metadata(Role),
-    {ok, Region} = fetch_document(),
+    % Fetch the IMDS session token and convert it to the request headers.
```
Comment (justincjohnson) [inline src/aws_credentials_ec2.erl:33]: This is the main part of the PR.  Getting the token first and passing it along in a header on every request.

File: src/aws_credentials_ec2.erl [hunk 5]
```diff
@@ -3,40 +3,70 @@
...
-fetch_role() ->
-    {ok, 200, Body, _Headers} = aws_credentials_httpc:get(?CREDENTIAL_URL),
+-spec fetch_session_token() -> {ok, session_token()}.
+fetch_session_token() ->
+  RequestHeaders = [{?SESSION_TOKEN_TTL_HEADER, ?SESSION_TOKEN_TTL_SECONDS}],
+  {ok, 200, Body, _Headers} = aws_credentials_httpc:put(?SESSION_TOKEN_URL, RequestHeaders),
     {ok, Body}.
 
-fetch_metadata(Role) ->
-    Url = lists:flatten(io_lib:format("~p/~p", [?CREDENTIAL_URL, Role])),
```
Comment (justincjohnson) [inline src/aws_credentials_ec2.erl:31]: This part right here is what was broken on master, from what I could tell.

File: src/aws_credentials_ec2.erl [hunk 6]
```diff
@@ -3,40 +3,70 @@
...
+  RequestHeaders = [{?SESSION_TOKEN_HEADER, SessionTokenString}],
+    {ok, RequestHeaders}.
+
+-spec fetch_role(request_headers()) -> {ok, role()}.
+fetch_role(RequestHeaders) ->
+    {ok, 200, Body, _Headers} = aws_credentials_httpc:get(?CREDENTIAL_URL, RequestHeaders),
+    {ok, Body}.
+
+fetch_metadata(Role, RequestHeaders) ->
+    Url = ?CREDENTIAL_URL ++ Role,
```
Comment (justincjohnson) [inline src/aws_credentials_ec2.erl:61]: Replaced the broken code with this.  But again, I'm new to Erlang so please educate me.

File: src/aws_credentials_httpc.erl [hunk 7]
```diff
@@ -10,20 +10,29 @@
...
+-export([start/0, get/1, get/2, get/3, put/1, put/2, put/3]).
 
 -define(PROFILE, aws_credentials).
 -define(TIMEOUT, 10000). % 10 sec
 -define(CONNECT_TIMEOUT, 3000). % 3 sec
 -define(DEFAULT_TRIES, 3).
+-define(DEFAULT_HEADERS, []).
 -define(DELAY, 100). % 100 microseconds
 
+-type field() :: string().
```
Comment (justincjohnson) [inline src/aws_credentials_httpc.erl:22]: Again, am I doing too much typing and spec'ing?

File: src/aws_credentials_httpc.erl [hunk 8]
```diff
@@ -10,20 +10,29 @@
...
+
 start() ->
     inets:start(httpc, [{profile, ?PROFILE}]).
 
 %% @doc Attempt to get a URL with the 3 retries. 3 is the default.
 get(URL) ->
-  get(URL, ?DEFAULT_TRIES).
+  get(URL, ?DEFAULT_HEADERS, ?DEFAULT_TRIES).
+
+get(URL, RequestHeaders) ->
```
Comment (justincjohnson) [inline src/aws_credentials_httpc.erl:34]: I changed get to allow request headers, and then essentially duplicated the `get` code for `put`.  It seems like I should combine these but thought I'd get feedback first.

File: src/aws_credentials_httpc.erl [hunk 9]
```diff
@@ -33,32 +42,72 @@ get(URL) ->
...
                                "~p of ~p) ~p: ~p.",
                                [NewRemaining, Tries, URL, Error]),
         timer:sleep((Tries - NewRemaining)*?DELAY),
-        get(URL, Tries, NewRemaining, [ Error | Errs ])
+        get(URL, RequestHeaders, Tries, NewRemaining, [ Error | Errs ])
     end.
 
-make_request(URL) ->
-    httpc:request(get, {URL, []},
+%% @doc Attempt to put a URL with the 3 retries. 3 is the default.
```
Comment (justincjohnson) [inline src/aws_credentials_httpc.erl:70]: All of the duped code.

File: src/aws_credentials_httpc.erl [hunk 10]
```diff
@@ -33,32 +42,72 @@ get(URL) ->
...
+
+      Error ->
+        NewRemaining = Remaining - 1,
+        error_logger:error_msg("Error fetching URL (attempts left: "
+                               "~p of ~p) ~p: ~p.",
+                               [NewRemaining, Tries, URL, Error]),
+        timer:sleep((Tries - NewRemaining)*?DELAY),
+        put(URL, RequestHeaders, Tries, NewRemaining, [ Error | Errs ])
+    end.
+make_request(Method, URL, RequestHeaders) ->
```
Comment (justincjohnson) [inline src/aws_credentials_httpc.erl:109]: Takes the `Method` now instead of hardcoding `get`.

File: src/aws_credentials_ec2.erl (hunk 2 above)
Comment (robertoaloi) [reply src/aws_credentials_ec2.erl:14]: Hi @justincjohnson, I believe you are right. This should be now solved in https://github.com/aws-beam/aws_credentials/pull/17 (just saw this PR), so it's probably a good idea to rebase this PR on top of the latest master to get that fix.

File: rebar.lock (hunk 1 above)
Comment (robertoaloi) [reply rebar.lock:1]: I have exactly the same bump at: https://github.com/aws-beam/aws_credentials/pull/19

File: src/aws_credentials_ec2.erl (hunk 3 above)
Comment (robertoaloi) [reply src/aws_credentials_ec2.erl:20]: I think you're doing great. This kind of "aliasing" for types improves readability in my opinion.

File: src/aws_credentials_ec2.erl (hunk 5 above)
Comment (robertoaloi) [reply src/aws_credentials_ec2.erl:31]: Indeed, it should be fixed now.

File: src/aws_credentials_ec2.erl (hunk 6 above)
Comment (justincjohnson) [reply src/aws_credentials_ec2.erl]: Rebased to pick up @robertoaloi's fix.

**Comment by robertoaloi** [comment]
Hi @justincjohnson Should we try to rebase this one on top of the new `master` and get this one in?

File: src/aws_credentials_httpc.erl (hunk 7 above)
Comment (mrallen1) [reply src/aws_credentials_httpc.erl:22]: No

File: src/aws_credentials_httpc.erl (hunk 8 above)
Comment (mrallen1) [reply src/aws_credentials_httpc.erl:34]: Ideally you'd have a `request` function that takes the method `get` or `put` as a parameter

**Review by mrallen1 — CHANGES_REQUESTED** [review]
In general this looks good- should convert this from a "draft" to a "full" PR. It's almost ready to merge. It would be super cool to get some tests for this new provider too, if you feel like adding them.

**Comment by justincjohnson** [comment]
I have finished rebasing off master and changing `get` and `put` functions to a single `request` function that takes the method as a parameter.

I will attempt to get some tests added now, once I get my test bearings :)

File: test/aws_credentials_providers_SUITE.erl [hunk 11]
```diff
@@ -141,12 +141,17 @@ teardown_provider(Context) ->
 
 mock_httpc_request_ec2(Method, Request, HTTPOptions, Options, Profile) ->
   case Request of
-    {"http://169.254.169.254/latest/meta-data/iam/security-credentials/", []} ->
+    {"http://169.254.169.254/latest/api/token/", _RequestHeaders} ->
+      {ok, response('session-token')};
+    % TODO: Pattern match the session token in RequestHeaders below
```
Comment (justincjohnson) [inline test/aws_credentials_providers_SUITE.erl:146]: There may be multiple headers in `RequestHeaders`.  I'd like to just ensure that the session token header returned from the first call appears in the appropriate header for subsequent requests.  Still looking into how to do this properly.

**Comment by justincjohnson** [comment]
I made some small changes to the tests so they pass.  You'll see I commented on `aws_credentials_providers_SUITE.erl` about the further improvements I was thinking of making.  This is all pretty minimal for testing but at least they pass now.  😬 

I also changed some lines to get past the linter errors on line length, quite possibly not in the style you would like, but I wasn't sure.

**Comment by justincjohnson** [comment]
Back to you @mrallen1.  Let me know if you want anything else done here.  I appreciate your time.

**Comment by justincjohnson** [comment]
Gentle reminder @mrallen1.  Thanks.

**Review by mrallen1 — APPROVED** [review]
lgtm

**Pull request merged**